physics/ephemeris_body.hpp

```diff
@@ -552,6 +555,115 @@ Status Ephemeris<Frame>::FlowWithAdaptiveStep(
   }
 }
 
+template<typename Frame>
+Status Ephemeris<Frame>::FlowWithAdaptiveStepGeneralized(
+      not_null<DiscreteTrajectory<Frame>*> trajectory,
+      GeneralIntrinsicAcceleration intrinsic_acceleration,
+      Instant const& t,
+      AdaptiveStepParameters const& parameters,
+      std::int64_t max_ephemeris_steps,
+      bool last_point_only) {
+  using ODE = ExplicitSecondOrderOrdinaryDifferentialEquation<Position<Frame>>;
+  Instant const& trajectory_last_time = trajectory->last().time();
+  if (trajectory_last_time == t) {
+    return Status::OK;
+  }
+
+  std::vector<not_null<DiscreteTrajectory<Frame>*>> const trajectories =
+      {trajectory};
+  // The |min| is here to prevent us from spending too much time computing the
+  // ephemeris.  The |max| is here to ensure that we always try to integrate
+  // forward.  We use |last_state_.time.value| because this is always finite,
+  // contrary to |t_max()|, which is -∞ when |empty()|.
+  Instant const t_final =
+      std::min(std::max(instance_time() +
+                            max_ephemeris_steps * parameters_.step(),
+                        trajectory_last_time + parameters_.step()),
+               t);
+  Prolong(t_final);
```

Comment:
I am sure that I have seen most of this code elsewhere.  Please factor it out.

Reply:
More TODOs for #1931.